gquant/dataframe_flow/task.py

```diff
@@ -62,14 +64,36 @@ def get_node_obj(self, replace=None):
                 spec.loader.exec_module(mod)
                 NodeClass = getattr(mod, node_type)
             else:
-                global MODLIB
+                global DEFAULT_MODULE
+                MODLIB = importlib.import_module(DEFAULT_MODULE)
                 NodeClass = getattr(MODLIB, node_type)
-        elif issubclass(node_type, Node):
+        elif issubclass(node_type, _Node):
             NodeClass = node_type
         else:
-            raise "Not supported"
+            raise Exception("Node type not supported: {}".format(node_type))
+
+        assert issubclass(NodeClass, _Node), \
+            'Node-type is not a subclass of "Node" class.'
+
+        if tgraph_mixin:
+            from ._node_flow import NodeTaskGraphMixin
+
+            class NodeInTaskGraph(NodeTaskGraphMixin, NodeClass):
+                def __repr__(self):
+                    '''Override repr to show the name and path of the plugin
+                    node class.'''
+                    return '<{} {}.{} object at {}>'.format(
+                        self.__class__.__name__,
+                        NodeClass.__module__,
+                        NodeClass.__name__,
+                        hex(id(self)))
+
+            node = NodeInTaskGraph(task)
+            # VERY IMPORTANT TO INIT NODE FOR TASKGRAPH PROCESSING
+            node._task_graph_init()
```

Comment:
The `_task_graph_init` is defined in the `NodeTaskGraphMixin`.
https://github.com/avolkov1/gQuant/blob/d32515c103fd29decdeaf47ca5bb071189c3adb5/gquant/dataframe_flow/_node_flow.py#L56

So far in the code there's one time when the  `NodeTaskGraphMixin` is not used, in the draw method:
https://github.com/avolkov1/gQuant/blob/d32515c103fd29decdeaf47ca5bb071189c3adb5/gquant/dataframe_flow/taskGraph.py#L214

The `_task_graph_init` is really meant to be used during task-graph build and run. I need the additional task-graph aware internal state variables added to a `Node` during task-graph run.
Build:
https://github.com/avolkov1/gQuant/blob/d32515c103fd29decdeaf47ca5bb071189c3adb5/gquant/dataframe_flow/taskGraph.py#L252
Run:
https://github.com/avolkov1/gQuant/blob/d32515c103fd29decdeaf47ca5bb071189c3adb5/gquant/dataframe_flow/taskGraph.py#L341

Although it would not hurt I didn't want to define `Node` class with the `NodeTaskGraphMixin`, because I wanted to keep internal state variables to a minimum and only introduce the mixin for when it's needed i.e. within the context of a task-graph build/run. So I dynamically define a `NodeInTaskGraph` class:
https://github.com/avolkov1/gQuant/blob/d32515c103fd29decdeaf47ca5bb071189c3adb5/gquant/dataframe_flow/task.py#L81

I will add a comment/docstring explaining `tgraph_mixin`.

Reply:
Does it make sense to include _task_graph_init inside the constructor of NodeInTaskGraph so it is initialized by default?

Reply:
I could do that. In python with multiple inheritance the `__init__` of other classes will not necessarily get called. You have to typically use `super` to explicitly call `__init__` of other classes. Currently the implementation looks like this:
```
            class NodeInTaskGraph(NodeTaskGraphMixin, NodeClass):
                def __repr__(self):
                    '''Override repr to show the name and path of the plugin
                    node class.'''
                    return '<{} {}.{} object at {}>'.format(
                        self.__class__.__name__,
                        NodeClass.__module__,
                        NodeClass.__name__,
                        hex(id(self)))

            node = NodeInTaskGraph(task)
            # VERY IMPORTANT TO INIT NODE FOR TASKGRAPH PROCESSING
            node._task_graph_init()
```

The only class with `__init__` is `Node` (`NodeClass` has to be a subclass of `Node` and should not have an `__init__` of its own, but if it does it should call super/`Node`'s init). Mixin classes in general should not have an `__init__` as they should not have data, only methods (<http://www.qtrac.eu/pyagg.html>) . Mixin classes are typically meant to be used as composable ojects with actions. But above actually the `NodeTaskGraphMixin` does introduce its own state/data. This state/data should be orthogonal to `Node` i.e. the `Node` class should know anything about data of `NodeTaskGraphMixin`.

If it makes more sense I could put the logic in `_task_graph_init` into an `__init__`. I wanted to be explicit with initialization of the task-graph relevant data. It would look like this using ` NodeTaskGraphMixin.__init__` instead:
```
            class NodeInTaskGraph(NodeTaskGraphMixin, NodeClass):
                def __init__(self, task):
                    NodeClass.__init__(self, task)
                    NodeTaskGraphMixin.__init__(self)

                def __repr__(self):
                    '''Override repr to show the name and path of the plugin
                    node class.'''
                    return '<{} {}.{} object at {}>'.format(
                        self.__class__.__name__,
                        NodeClass.__module__,
                        NodeClass.__name__,
                        hex(id(self)))

            node = NodeInTaskGraph(task)
```

Reply:
Changed to use `__init__` instead of `_task_graph_init`.